File: Godot/WindowPlayerController.cs
using Godot;

namespace Game5;

public class WindowPlayerController : IWindowPlayerController
{
	private const float Speed = 5.0f;
	private const float JumpImpulse = 5.0f;
	private const float MouseSensitivity = 0.01f;
	private const float RollSensitivity = 0.01f;
	private readonly WindowPlayer _windowPlayer;

	public WindowPlayerController(WindowPlayer windowPlayer)
	{
		_windowPlayer = windowPlayer;
	}
	
	public void ProcessMovementControls()
	{
		var velocity = _windowPlayer.Velocity;
		var cameraPivot = _windowPlayer.GetNode<Marker3D>("CameraPivot");

		var inputDir = Input.GetVector("move_left", "move_right", 
			"move_forward", "move_back");
		var yMovement = _windowPlayer.IsMaterialized() ? 0 : 
			Input.GetAxis("move_down", "move_up");
		var direction = (_windowPlayer.Transform.Basis * 
						 new Vector3(inputDir.X, yMovement, inputDir.Y)).Normalized();
		if (direction != Vector3.Zero)
		{
			velocity.X = direction.X * Speed;
			velocity.Z = direction.Z * Speed;
			if (!_windowPlayer.IsMaterialized())
			{
				velocity.Y = direction.Y * Speed;
			}
		}
		else
		{
			velocity.X = Mathf.MoveToward(velocity.X, 0, Speed);
			velocity.Z = Mathf.MoveToward(velocity.Z, 0, Speed);
			if (!_windowPlayer.IsMaterialized())
			{
				velocity.Y = Mathf.MoveToward(velocity.Y, 0, Speed);
			}
		}
		if (_windowPlayer.IsMaterialized() && _windowPlayer.IsOnFloor() && Input.IsActionJustPressed("move_up"))
		{
			velocity.Y = JumpImpulse;
		}

		_windowPlayer.Velocity = velocity;

		var cameraX = Input.GetAxis("look_down", "look_up");
		var cameraY = Input.GetAxis("look_left", "look_right");
		var cameraZ = Input.GetAxis("roll_left", "roll_right");
		_windowPlayer.RotateY(-cameraY * MouseSensitivity);
		cameraPivot.RotateX(cameraX * MouseSensitivity);
		cameraPivot.RotateZ(-cameraZ * RollSensitivity);
	}

	public void ProcessMouseControls(InputEvent @event)
	{
		
		if (((!_windowPlayer.IsMaterialized() && Input.IsActionPressed("holding_right_mouse_button"))
			 || (_windowPlayer.IsMaterialized() && Input.MouseMode == Input.MouseModeEnum.Captured))
			&& @event is InputEventMouseMotion eventMouseMotion)
		{
			_windowPlayer.RotateY(-eventMouseMotion.Relative.X * MouseSensitivity);
			_windowPlayer.GetNode<Marker3D>("CameraPivot")
				.RotateX(-eventMouseMotion.Relative.Y * MouseSensitivity);
		}

		if (_windowPlayer.IsMaterialized())
		{
			if (Input.IsActionPressed("click") && Input.MouseMode == Input.MouseModeEnum.Visible)
			{
				Input.MouseMode = Input.MouseModeEnum.Captured;
			}

			if (Input.IsActionPressed("ui_cancel"))
			{
				Input.MouseMode = Input.MouseModeEnum.Visible;
			}
		}
	}
}

File: Sources/Game/WindowPlayerController.swift
//
//  WindowPlayerController.swift
//
//
//  Created by Anthony Michalek on 2/18/24.
//

import SwiftGodot;

class WindowPlayerController: WindowPlayerControllerProtocol {
    private final var speed: Float = 5.0;
    private final var friction: Double = 0.1;
    private final var jumpImpulse: Float = 5.0;
    private final var mouseSensitivity: Double = 0.01;
    private final var rollSensitivity: Double = 0.01;
    private final var windowPlayer: WindowPlayer;
    
    init(windowPlayer: WindowPlayer)
    {
        self.windowPlayer = windowPlayer;
    }
    
    func processMovementControls() {
        var cameraPivot: Marker3D = windowPlayer.getNode(path: "CameraPivot") as! Marker3D;
        
        var inputDir: Vector2 = Input.getVector(negativeX: "move_left", positiveX: "move_right",
         negativeY: "move_forward", positiveY: "move_back");
        var yMovement: Float = windowPlayer.isMaterialized() ? 0 :
        Float(Input.getAxis(negativeAction: "move_down", positiveAction: "move_up"));
        var direction: Vector3 = (windowPlayer.transform.basis * Vector3(x: inputDir.x, y: yMovement, z: inputDir.y));
        if (direction != .zero)
        {
            windowPlayer.velocity.x = direction.x * speed;
            windowPlayer.velocity.z = direction.z * speed;
            if (!windowPlayer.isMaterialized())
            {
                windowPlayer.velocity.y = direction.y * speed;
            }
        }
        else
        {
            windowPlayer.velocity.moveToward(to: Vector3(x: 0.0, y: yMovement, z: 0.0), delta: friction);
        }
        if (windowPlayer.isMaterialized() && windowPlayer.isOnFloor() && Input.isActionJustPressed(action: "move_up"))
        {
            windowPlayer.velocity.y = jumpImpulse;
        }
        
        var cameraX = Input.getAxis(negativeAction: "look_down", positiveAction: "look_up");
        var cameraY = Input.getAxis(negativeAction: "look_left", positiveAction: "look_right");
        var cameraZ = Input.getAxis(negativeAction: "roll_left", positiveAction: "roll_right");
        windowPlayer.rotateY(angle: -cameraY * mouseSensitivity);
        cameraPivot.rotateX(angle: cameraX * mouseSensitivity);
        cameraPivot.rotateZ(angle: -cameraZ * rollSensitivity);
    }
    
    func processMouseControls(event: InputEvent) {
        if (((!windowPlayer.isMaterialized() && Input.isActionPressed(action: "holding_right_mouse_button"))
             || (windowPlayer.isMaterialized() && Input.mouseMode == Input.MouseMode.captured))
            && event is InputEventMouseMotion)
        {
            windowPlayer.rotateY(angle: -Double((event as! InputEventMouseMotion).relative.x) * mouseSensitivity);
            (windowPlayer.getNode(path: "CameraPivot") as! Marker3D).rotateX(angle:
            -Double((event as! InputEventMouseMotion).relative.y) * mouseSensitivity)
        }
        
        if (windowPlayer.isMaterialized())
        {
            if (Input.isActionPressed(action: "click") && Input.mouseMode == Input.MouseMode.visible)
            {
                Input.mouseMode = Input.MouseMode.captured;
            }
            
            if (Input.isActionPressed(action: "ui_cancel"))
            {
                Input.mouseMode = Input.MouseMode.visible;
            }
        }
    }
    
}
